### ISSUE TITLE
Disambiguate "George W. Bush" and "George H. W. Bush"

### ISSUE THREAD

**Issue by loretoparisi** [facebookresearch/GENRE#42]
To correctly disambiguate the two US presidents, "George W. Bush" and "George H. W. Bush"
I have tried several approaches using `hf_e2e_entity_linking_aidayago` and `hf_e2e_entity_linking_wiki_abs` models:

```python
# Example: End-to-End Entity Linking
# wikipedia aidayago
model = GENRE.from_pretrained(os.path.join(cache_dir,"hf_e2e_entity_linking_aidayago")).eval()
# or wikipedia
wiki_model = GENRE.from_pretrained(os.path.join(cache_dir,"hf_e2e_entity_linking_wiki_abs")).eval()
```

w/ mention_trie, mention_to_candidates_dict
```python
sentences = ["George Bush was the 43rd president of United States"]
entity_spans = get_entity_spans(
    model,
    sentences,
    mention_trie=Trie([
        model.encode(" {}".format(e))[1:].tolist()
        for e in ["George Bush"]
    ]),
     mention_to_candidates_dict={
        "George Bush": ["George W. Bush", "George H. W. Bush"]
    }
)
print(get_markdown(sentences, entity_spans)[0])
```

Result: **WRONG**  👎🏾 
```
[George Bush](https://en.wikipedia.org/wiki/George_H._W._Bush) was the 43rd president of United States
```

```python
sentences = ["George Bush was the 43rd president of United States"]
entity_spans = get_entity_spans(
    model,
    sentences,
    mention_trie=Trie([
        model.encode(" {}".format(e))[1:].tolist()
        for e in ["George Bush"]
    ]),
    mention_to_candidates_dict={
        "George Bush": ["George W. Bush"]
    }
)
print(get_markdown(sentences, entity_spans)[0])
```

Result: **CORRECT** 👍🏾 🥇 
```
[George Bush](https://en.wikipedia.org/wiki/George_W._Bush) was the 43rd president of United States
```

w/ candidates_trie
```python
sentences = ["George Bush was the 43rd president of the United States from 2001 to 2009"]

prefix_allowed_tokens_fn = get_prefix_allowed_tokens_fn(
    model,
    sentences,
    candidates_trie=Trie([
        model.encode(" }} [ {} ]".format(e))[1:].tolist()
        for e in ["George W. Bush", "George H. W. Bush"]
    ])
)
out = model.sample(
    sentences,
    prefix_allowed_tokens_fn=prefix_allowed_tokens_fn,
)
print(out)
```
Result:WRONG (+ possible bug...) 👎🏾 
```
[[{'text': 'George { Bush } [ George H. W. Bush ] was the 43rd president of the United States from 2001 to 2009', 'logprob': tensor(-0.7810)}], [{'text': 'George { Bush } [ George H. W. Bush ] was the 43rd president of the United States from 2001 to { 2009 } [ and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and ...
```

but using 

```python
sentences = ["George Bush was the 41rd president of the United States from 1989 to 2003"]
entity_spans = get_entity_spans(
    wiki_model,
    sentences,
    mention_trie=Trie([
        model.encode(" {}".format(e))[1:].tolist()
        for e in ["George Bush"]
    ]),
    mention_to_candidates_dict={
        "George Bush": ["George W. Bush", "George H. W. Bush"]
    }
)
print(get_markdown(sentences, entity_spans)[0])
```

CORRECT Result: 👍🏾  🥇 
```
[George Bush](https://en.wikipedia.org/wiki/George_H._W._Bush) was the 41rd president of the United States from 2001 to 2009
```

and

```python
sentences = ["George Bush was the 43rd president of the United States from 2001 to 2009"]
entity_spans = get_entity_spans(
    wiki_model,
    sentences,
    mention_trie=Trie([
        model.encode(" {}".format(e))[1:].tolist()
        for e in ["George Bush"]
    ]),
    mention_to_candidates_dict={
        "George Bush": ["George W. Bush", "George H. W. Bush"]
    }
)
print(get_markdown(sentences, entity_spans)[0])
```

WRONG Result: 👎🏾 
```
[George Bush](https://en.wikipedia.org/wiki/George_H._W._Bush) was the 43rd president of the United States from 2001 to 2009
```

I therefore approached with kilt wikipedia trie and `hf_entity_disambiguation_aidayago` model

```python
with open(os.path.join(cache_dir,"kilt_titles_trie_dict.pkl"), "rb") as f:
    trie = Trie.load_from_dict(pickle.load(f))
dmodel = GENRE.from_pretrained(os.path.join(cache_dir,"hf_entity_disambiguation_aidayago")).eval()
sentences = ["[START_ENT] George Bush [END_ENT] was the 43rd president of United States"]
out = dmodel.sample(
    sentences,
    prefix_allowed_tokens_fn=lambda batch_id, sent: trie.get(sent.tolist()),
)
print(out)
```

WRONG Result 👎🏾 
```
[[{'text': 'George H. W. Bush', 'logprob': tensor(-0.0866)}], [{'text': 'George W. Bush', 'logprob': tensor(-0.7922)}], [{'text': 'George H. W. Bush vomiting incident', 'logprob': tensor(-1.6464)}], [{'text': 'George H. W. Bush 1988 presidential campaign', 'logprob': tensor(-1.6795)}], [{'text': 'George H. W. Bush Supreme Court candidates', 'logprob': tensor(-2.1032)}]]
```

and for the model `hf_wikipage_retrieval`:

```python
smodel = GENRE.from_pretrained(os.path.join(cache_dir,"hf_wikipage_retrieval")).eval()
sentences = ["[START_ENT] George Bush [END_ENT] was the 43rd president of United States"]
out = smodel.sample(
    sentences,
    prefix_allowed_tokens_fn=lambda batch_id, sent: trie.get(sent.tolist()),
)
print(out)
```

WRONG Result 👎🏾 
```
[[{'text': 'George H. W. Bush', 'logprob': tensor(-0.0850)}], [{'text': 'George W. Bush', 'logprob': tensor(-0.7715)}], [{'text': 'George H. W. Bush Supreme Court candidates', 'logprob': tensor(-1.3834)}], [{'text': 'George H. W. Bush 1988 presidential campaign', 'logprob': tensor(-1.4211)}], [{'text': 'George H. W. Bush vomiting incident', 'logprob': tensor(-2.1070)}]]
```

So to recap, I have got only two cases when the best prediction is the desired one:

```python
sentences = ["George Bush was the 43rd president of United States"]
entity_spans = get_entity_spans(
    model,
    sentences,
    mention_trie=Trie([
        model.encode(" {}".format(e))[1:].tolist()
        for e in ["George Bush"]
    ]),
    mention_to_candidates_dict={
        "George Bush": ["George W. Bush"]
    }
)
print(get_markdown(sentences, entity_spans)[0])
```

Result: **CORRECT** 👍🏾 🥇 
```
[George Bush](https://en.wikipedia.org/wiki/George_W._Bush) was the 43rd president of United States
```

and 

```python
sentences = ["George Bush was the 41rd president of the United States from 1989 to 2003"]
entity_spans = get_entity_spans(
    wiki_model,
    sentences,
    mention_trie=Trie([
        model.encode(" {}".format(e))[1:].tolist()
        for e in ["George Bush"]
    ]),
    mention_to_candidates_dict={
        "George Bush": ["George W. Bush", "George H. W. Bush"]
    }
)
print(get_markdown(sentences, entity_spans)[0])
```

CORRECT Result: 👍🏾  🥇 
```
[George Bush](https://en.wikipedia.org/wiki/George_H._W._Bush) was the 41rd president of the United States from 2001 to 2009
```

Questions
- Am I missing some other ways to disambiguate?
- Assumed that both [George W. Bush](https://it.wikipedia.org/wiki/George_W._Bush) and [George H. W. Bush](https://it.wikipedia.org/wiki/George_H._W._Bush) are in the knowledge graph, do the model needs more additional context like other `mention_to_candidates_dict` or `mention_trie ` possibile values?
- Supposed that one president (like `George W. Bush`) was missing, how to add his wikipedia entity page?
- Are the "wrong" cases formally correct or am I missing something?

Thanks a lot!

**Comment by paulthemagno** [comment]
Very interesting issue... +1

**Comment by nicola-decao** [comment]
Hi @loretoparisi your comment is quite hard to parse. Do you think you spotted a bug or are you just pointing to model mistakes? If the model makes mistakes there is nothing I can do since it just converged to that solution

**Comment by loretoparisi** [comment]
@nicola-decao thanks not sure if one could file it "as a bug" or just the prediction to which the model converges to (FP or TP, depending on the input as we can see above).
But if we look at those two adversary results, we can argue that, for a given 

- mention tree

`mention_trie=Trie([model.encode(" {}".format(e))[1:].tolist() for e in ["George Bush"]])`

- mention to candidates

`mention_to_candidates_dict={"George Bush": ["George W. Bush", "George H. W. Bush"]}`

We have two opposite results for the two "contrastive" sentences 

`sentences = ["George Bush was the 41rd president of the United States from 1989 to 2003"]`

That works and the latter that instead fails 

`sentences = ["George Bush was the 43rd president of the United States from 2001 to 2009"]
`

Despite of the fact that we have both the candidates in the mention to candidates mappings:


`{ "George Bush": ["George W. Bush", "George H. W. Bush"]}`

So it seems to me to be a bug (sort of), but - from the model's perspective, maybe it is not, and it's just a prediction error.
If it's not a bug, the my question is: how to mitigate that coverging-to-false-positive effect?

Thanks a lot!

> Hi @loretoparisi your comment is quite hard to parse. Do you think you spotted a bug or are you just pointing to model mistakes? If the model makes mistakes there is nothing I can do since it just converged to that solution
> 
>

**Comment by nicola-decao** [comment]
For me, a **bug** is a mistake we made in the code where a **prediction error** is a mistake from the model. Let's remember that my model has not accuracy 100% and I haven't solved EL.  Here I can only see that the model make a **prediction error**  so there is nothing I can fix.

Regarding the mitigation of such errors: if you consistently find that the model is confused between the two presidents you can simply find some datapoints and fine-tune on them. I cannot think of other solution right now.

**Comment by loretoparisi** [comment]
@nicola-decao yes agreed, in fact that was my doubt. Could point me to the way to provide a new training set and fine-tune? Thanks a lot!

**Comment by nicola-decao** [comment]
One way would be asking annotators in MTurk to write or retrieve from the web sentences about George W. Bush and George H. W. Bush and disambiguate

**Comment by paulthemagno** [comment]
@nicola-decao I encountered the same problem (I think a **bug** in this case) of @loretoparisi in that example: 

> w/ candidates_trie
> 
> ```python
> sentences = ["George Bush was the 43rd president of the United States from 2001 to 2009"]
> 
> prefix_allowed_tokens_fn = get_prefix_allowed_tokens_fn(
>     model,
>     sentences,
>     candidates_trie=Trie([
>         model.encode(" }} [ {} ]".format(e))[1:].tolist()
>         for e in ["George W. Bush", "George H. W. Bush"]
>     ])
> )
> out = model.sample(
>     sentences,
>     prefix_allowed_tokens_fn=prefix_allowed_tokens_fn,
> )
> print(out)
> ```
> 
> Result:WRONG (+ possible bug...) 👎🏾
> 
> ```
> [[{'text': 'George { Bush } [ George H. W. Bush ] was the 43rd president of the United States from 2001 to 2009', 'logprob': tensor(-0.7810)}], [{'text': 'George { Bush } [ George H. W. Bush ] was the 43rd president of the United States from 2001 to { 2009 } [ and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and and ...
> ```

In this case `2009` is taken as an entity but the result is a series of `and`: `[ and and and and and and and and and and and and...]`. Is this behavior expected? If you prefer I can open a new issue about it.
